src/ValueObjects/Age.php

```diff
@@ -20,8 +23,8 @@ protected static function getDecoder(): Decoder
     {
         return Decoders::pipe(
             Decoders::int(),
-            /** @psalm-var Decoder<int, self> */
             Decoders::make(
+                /** @return Validation<self> */
```

Comment:
This is the part that I do not like, you have to fix this manually every time.
Maybe this is due to `Validation::failure` that says `@psalm-return ValidationFailures<T>` instead of `@psalm-return ValidationFailures<never>`?

Reply:
I agree on disliking that part.
Looking further, what i see is that the codomain of `Decoders::make` is a `Validation<A>` where `A` is inferred by the "sum" of the codomain of the `Validation::success` and `Validation::failure`.
I guess that the blame is on the `Validation::failure` sign that has no clue to infer the right type of what is failed to validates here.

A possibile suggested solution is to tight the type of `A` inferred from the failure method making it explicit. 
```php
// Validation.php

...
/**
     * @psalm-template T
     * @psalm-param mixed $value
     * @psalm-param Context $context
     * @psalm-param class-string<T> $of
     * @psalm-param string|null $message
     * @psalm-return ValidationFailures<T>
     *
     * @param mixed $value
     */
    public static function failureT(string $of, $value, Context $context, ?string $message = null): self
    {
        return self::failures(
            [new VError($value, $context, $message)]
        );
    }
...
```

However, this requires you to make changes to the `php-codec` library. 
I have no clue on how to solve it purely on client side.

Reply:
IMHO that's not possibile, even on the lib side: `A` could be not existing in a case of failure, especially in our case: the VO shouldn't exist in an invalid state, but in that way you're forcing me to create it as invalid.

Maybe you could force the failure to contain a `class-string<A>`, but it seems clunky.

@ilario-pierbattista WDYT?

Reply:
I see your point, and I am aware of this issue.

My goal is to have `Validation<T>` being isomorphic to a generic `Either<VError[]. T>`.
Maybe I could find a better implementation in other function libraries, such as https://github.com/marcosh/lamphpda .

Btw, `class-string<T>` won't work in all cases, since it needs to deal also with non-object values.

Reply:
https://github.com/facile-it/php-codec/issues/36